[SPARK-9384] [core] Easier setting of executor and driver classpath

markgrover opened the pull request:
Adding new property spark.common.extraClassPath for adding to both the driver and executor classpath.

More testing needs to be done.

Currently one can set extra jars in the driver classpath using spark.driver.extraClassPath property. Similarly, the executor's extra classpath can be set using spark.executor.extraClassPath property.
However, many users (myself) included have often suffered from the dogma of setting one and missing the other. Causing a lot of wasted time. And, this JIRA is an effort to see if the status quo can be made better.

--jars sets the classpath set by the classloader but there is no such way of setting the system classpath of the driver and the executor. This change would allow users to override libraries shipped with Spark, or add jars like some JDBC driver, that don't work unless they are adding to the system classpath.

Testing done:
1. Updated existing unit tests, added a few more, especially for the driverClassPath. If someone has thoughts on where adding tests for testing the executor classpath would make sense, I'd love to hear.
2. Tested this on a real cluster with one master and one executor, running Spark in standalone mode. Verified that both the spark driver and executor classpath were being prepended to.
3. Documentation in the github repo updated as well.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

squito commented:
Jenkins, ok to test

sarutak commented:
ok to test.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38832 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38832/consoleFull) for   PR 7739 at commit [`5173f99`](https://github.com/apache/spark/commit/5173f9910b12e5446f76d6c6fe96a12ddc201254).

SparkQA commented:
[Test build #38832 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38832/console) for   PR 7739 at commit [`5173f99`](https://github.com/apache/spark/commit/5173f9910b12e5446f76d6c6fe96a12ddc201254).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

markgrover commented:
I will take a look at the failures.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38871 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38871/consoleFull) for   PR 7739 at commit [`62498c3`](https://github.com/apache/spark/commit/62498c34af16f9bce6738ed8dbfbd5f72c5d7b45).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38876 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38876/consoleFull) for   PR 7739 at commit [`0717f16`](https://github.com/apache/spark/commit/0717f1636e24c067064484fc8b0e7b39a352b884).

SparkQA commented:
[Test build #38876 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38876/console) for   PR 7739 at commit [`0717f16`](https://github.com/apache/spark/commit/0717f1636e24c067064484fc8b0e7b39a352b884).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shaneknapp commented:
sorry, i broke the jenkins config and have since reverted my change.  allow me to retrigger your build...  :)

jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38871 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38871/console) for   PR 7739 at commit [`62498c3`](https://github.com/apache/spark/commit/62498c34af16f9bce6738ed8dbfbd5f72c5d7b45).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

markgrover commented:
Still looks like a file missing in Jenkins workspace:
```
13:55:21.583 ERROR org.apache.spark.executor.Executor: Exception in task 0.0 in stage 11926.0 (TID 84068)
java.io.FileNotFoundException: File file:/home/jenkins/workspace/SparkPullRequestBuilder@3/target/tmp/spark-ce042421-1af0-42fd-a04a-892d72b3df72/tmp/part-00003 does not exist
```

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39294 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39294/consoleFull) for   PR 7739 at commit [`51a98b4`](https://github.com/apache/spark/commit/51a98b45988ba37311023ea2d4f3f6af1ed0ce7c).

SparkQA commented:
[Test build #39294 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39294/console) for   PR 7739 at commit [`51a98b4`](https://github.com/apache/spark/commit/51a98b45988ba37311023ea2d4f3f6af1ed0ce7c).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #39398 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39398/consoleFull) for   PR 7739 at commit [`d905b92`](https://github.com/apache/spark/commit/d905b92329c852b8b0ea66f59612d9d5adc6bd2e).

SparkQA commented:
[Test build #39398 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39398/console) for   PR 7739 at commit [`d905b92`](https://github.com/apache/spark/commit/d905b92329c852b8b0ea66f59612d9d5adc6bd2e).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #39591 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39591/consoleFull) for   PR 7739 at commit [`8ebf8e1`](https://github.com/apache/spark/commit/8ebf8e1d0f4cb4f75f5ea5b77b9c72feb7b19207).

SparkQA commented:
[Test build #39591 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39591/console) for   PR 7739 at commit [`8ebf8e1`](https://github.com/apache/spark/commit/8ebf8e1d0f4cb4f75f5ea5b77b9c72feb7b19207).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #39598 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39598/consoleFull) for   PR 7739 at commit [`163e8f6`](https://github.com/apache/spark/commit/163e8f68595884947220b7e9ccbb96e56a982d1c).

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #39603 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39603/consoleFull) for   PR 7739 at commit [`d81bfa3`](https://github.com/apache/spark/commit/d81bfa33e7c20d33b7c0cb7955530e68b79d38b1).

SparkQA commented:
[Test build #39598 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39598/console) for   PR 7739 at commit [`163e8f6`](https://github.com/apache/spark/commit/163e8f68595884947220b7e9ccbb96e56a982d1c).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

SparkQA commented:
[Test build #39603 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39603/console) for   PR 7739 at commit [`d81bfa3`](https://github.com/apache/spark/commit/d81bfa33e7c20d33b7c0cb7955530e68b79d38b1).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

sarutak commented:
Hey @markgrover , could you rebase your local repository and then apply your change?
The lineage is messy.

markgrover commented:
Sure thing. Will do!
On Aug 3, 2015 10:18 PM, "Kousuke Saruta" <notifications@github.com> wrote:

> Hey @markgrover <https://github.com/markgrover> , could you rebase your
> local repository and then apply your change?
> The lineage is messy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7739#issuecomment-127482044>.
>

markgrover committed:
Adding new property spark.common.extraClassPath for adding to both the driver and executor classpath.

Currently one can set extra jars in the driver classpath using spark.driver.extraClassPath property. Similarly, the executor's extra classpath can be set using spark.executor.extraClassPath property.
However, many users (myself) included have often suffered from the dogma of setting one and missing the other. Causing a lot of wasted time. And, this JIRA is an effort to see if the status quo can be made better.

--jars sets the classpath set by the classloader but there is no such way of setting the system classpath of the driver and the executor. This change would allow users to override libraries shipped with Spark, or add jars like some JDBC driver, that don't work unless they are adding to the system classpath.

markgrover commented:
Hi @sarutak, thanks again for taking a look, I've rebased and fixed up the history. I have also update the PR with some more info.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39716 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39716/consoleFull) for   PR 7739 at commit [`e33a404`](https://github.com/apache/spark/commit/e33a404e586fb992ad9de7ef751358385ba2a4a9).

SparkQA commented:
[Test build #39716 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39716/console) for   PR 7739 at commit [`e33a404`](https://github.com/apache/spark/commit/e33a404e586fb992ad9de7ef751358385ba2a4a9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

markgrover commented:
Seems unrelated failure. I don't think I am on the whitelist but I will try this:

Jenkins, test this please.

markgrover commented:
If that doesn't work and some can retrigger the test run on my behalf, I'd appreciate that. cc @squito @sarutak

squito commented:
Jenkins, add to whitelist

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build triggered.

markgrover commented:
Thanks @squito !

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #231 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/231/consoleFull) for   PR 7739 at commit [`e33a404`](https://github.com/apache/spark/commit/e33a404e586fb992ad9de7ef751358385ba2a4a9).

SparkQA commented:
[Test build #39860 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39860/consoleFull) for   PR 7739 at commit [`e33a404`](https://github.com/apache/spark/commit/e33a404e586fb992ad9de7ef751358385ba2a4a9).

SparkQA commented:
[Test build #231 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/231/console) for   PR 7739 at commit [`e33a404`](https://github.com/apache/spark/commit/e33a404e586fb992ad9de7ef751358385ba2a4a9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #39860 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39860/console) for   PR 7739 at commit [`e33a404`](https://github.com/apache/spark/commit/e33a404e586fb992ad9de7ef751358385ba2a4a9).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40171 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40171/consoleFull) for   PR 7739 at commit [`e33a404`](https://github.com/apache/spark/commit/e33a404e586fb992ad9de7ef751358385ba2a4a9).

SparkQA commented:
[Test build #40171 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40171/console) for   PR 7739 at commit [`e33a404`](https://github.com/apache/spark/commit/e33a404e586fb992ad9de7ef751358385ba2a4a9).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

markgrover commented:
I am taking a look.

markgrover commented:
Hmm, looks like PYTHONPATH is not being set up properly by some python tests. It turns out I am not the only suffering from this - the build before mine (https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40170/console) failed because of the same reason. I am working on something else at the moment but if it's not fixed by the time I get back to this, I will take a look at the orthogonal cause that's causing all these test builds to fail.

markgrover commented:
Jenkins, retest this please.

harishreedharan commented on core/src/main/scala/org/apache/spark/deploy/Client.scala:
```diff
@@ -68,7 +68,12 @@ private class ClientEndpoint(
         val mainClass = "org.apache.spark.deploy.worker.DriverWrapper"
 
         val classPathConf = "spark.driver.extraClassPath"
```
Rename this to `driverClasspathConf` ?

markgrover replied:
ok, will do.

harishreedharan commented on core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolRequest.scala:
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.deploy.rest
 
+import java.io.File
```
Why?

markgrover replied:
Sorry, will remove.

harishreedharan commented on core/src/main/scala/org/apache/spark/deploy/rest/StandaloneRestServer.scala:
```diff
@@ -135,7 +135,10 @@ private[rest] class StandaloneSubmitRequestServlet(
-    val driverExtraClassPath = sparkProperties.get("spark.driver.extraClassPath")
+    val commonExtraClassPath = sparkProperties.get("spark.common.extraClassPath")
+    val driverSpecificExtraClassPath = sparkProperties.get("spark.driver.extraClassPath")
+    val driverExtraClassPath = Option(( commonExtraClassPath ++ driverSpecificExtraClassPath)
```
Nit: unnecessary space after `((`

harishreedharan replied:
Ah ok, they can't be nulls since `get` returns an option (I thought it was a Java map)

markgrover replied:
Will remove the space.

harishreedharan commented on core/src/main/scala/org/apache/spark/deploy/rest/mesos/MesosRestServer.scala:
```diff
@@ -88,7 +88,10 @@ private[mesos] class MesosSubmitRequestServlet(
-    val driverExtraClassPath = sparkProperties.get("spark.driver.extraClassPath")
+    val commonExtraClassPath = sparkProperties.get("spark.common.extraClassPath")
+    val driverSpecificExtraClassPath = sparkProperties.get("spark.driver.extraClassPath")
+    val driverExtraClassPath = Option(( commonExtraClassPath ++ driverSpecificExtraClassPath)
```
Nit: Unneeded space after `((`

markgrover replied:
Will fix.

harishreedharan commented on core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala:
```diff
@@ -61,8 +61,10 @@ private[spark] class SparkDeploySchedulerBackend(
-    val classPathEntries = sc.conf.getOption("spark.executor.extraClassPath")
-      .map(_.split(java.io.File.pathSeparator).toSeq).getOrElse(Nil)
+
+    val classPathEntries = sc.conf.getExecutorExtraClassPath().map(_.split(java.io.File
```
Formatting is a little off. Can you split this into 2 lines a bit better? Or you could:
`val classPathEntries = 
  sc.conf.getExecutorExtraClassPath()
  .map(_.split(java.io.File.pathSeparator).toSeq).getOrElse(Nil)`

markgrover replied:
Will do.

harishreedharan commented:
This looks pretty good. Could you also add some tests to each of the components (YARN/standalone and mesos to ensure the correct classpaths get passed in the correct order?)

squito commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -479,6 +484,23 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
+    // Concatenate common and driver classpath in that order
+    val commonExtraClassPath = getOption("spark.common.extraClassPath")
+    val driverExtraClassPath = getOption("spark.driver.extraClassPath")
+    Option((commonExtraClassPath ++ driverExtraClassPath).mkString(File.pathSeparator)).filter(_
```
this is super minor, but using `Option` as a wrapper makes the reader think that the you are checking for whether or not there is null inside.

```scala
scala> Option("foo")
res20: Option[String] = Some(foo)

scala> Option("")
res21: Option[String] = Some()

scala> Option(null)
res22: Option[Null] = None
```

but really, you know there won't be any nulls, you are just trying to wrap the String in an option.  The more conventional thing to do in that case is to wrap with `Some`.  The end result is the same in this case -- both do the same thing when the input isn't null -- but the intention is a bit more clear.

markgrover replied:
Ok, will consider.

markgrover replied:
Actually, it can be null, right? If neither spark.common.extraClassPath nor spark.driver.extraClassPath are set, I'd like this function to return None so I am not changing the original semantics of anybody calling this function. And, if so, casting as option is right thing, no?

If there is a more legible way of expressing this, I am totally open to it.

markgrover replied:
Hmm, looks like it doesn't matter whether I use Some or None.
````
scala> Option((None++None).mkString(":")).filter(_.nonEmpty)
res3: Option[String] = None

scala> Some((None++None).mkString(":")).filter(_.nonEmpty)
res4: Option[String] = None
```

squito commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala:
```diff
@@ -88,7 +88,10 @@ private[spark] class MesosSchedulerBackend(
+
+    val userClassPath = sc.conf.getExecutorExtraClassPath()
+
+    userClassPath.foreach { cp =>
```
nit: no real benefit to the tmp variable, you can keep it like before `sc.conf.getExecutorExtraClassPath().foreach { ...`

markgrover replied:
Will do, thanks.

markgrover commented:
Thanks @harishreedharan and @squito for your reviews!

markgrover committed:
Incorporating feedback from the review

markgrover committed:
Minor formatting changing. Changing it back to Option instead of Some()

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #41272 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41272/consoleFull) for   PR 7739 at commit [`8740751`](https://github.com/apache/spark/commit/87407517c333034041c4312a76b0641dafe4fd2e).

markgrover commented:
Hey @harishreedharan about
> Could you also add some tests to each of the components (YARN/standalone and mesos to ensure the correct classpaths get passed in the correct order?)
I added some tests with the original pull request. Could you please help me better understand where exactly in the code you'd like me to add new tests? Thanks!

markgrover committed:
Converting Option cast to Some for clarity

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

markgrover committed:
Updating branch with latest code from master. Resolving merge conflict

SparkQA commented:
[Test build #41275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41275/consoleFull) for   PR 7739 at commit [`5f62d39`](https://github.com/apache/spark/commit/5f62d39aadafc1780570c368fe6ceeaa39186758).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41277/consoleFull) for   PR 7739 at commit [`bca4fa3`](https://github.com/apache/spark/commit/bca4fa303a6de1b8b395c9fc6c2357edb9e12d5e).

SparkQA commented:
[Test build #41272 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41272/console) for   PR 7739 at commit [`8740751`](https://github.com/apache/spark/commit/87407517c333034041c4312a76b0641dafe4fd2e).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41272/
Test FAILed.

SparkQA commented:
[Test build #41275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41275/console) for   PR 7739 at commit [`5f62d39`](https://github.com/apache/spark/commit/5f62d39aadafc1780570c368fe6ceeaa39186758).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41275/
Test PASSed.

SparkQA commented:
**[Test build #41277 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41277/console)**     for PR 7739 at commit [`bca4fa3`](https://github.com/apache/spark/commit/bca4fa303a6de1b8b395c9fc6c2357edb9e12d5e)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41277/
Test FAILed.

markgrover commented:
Jenkins, test this please.

harishreedharan commented:
Since most of the changes don't affect the cluster managers directly, and spark submit changes have tests, I think this is good (I was confused by the one change in `YarnClusterSuite`).

markgrover commented:
Thanks Hari! The YarnClusterSuite change is rather benign and not absolutely necessary. So, based on this I believe this request is ready to be merged.

And, for some reason, Jenkins is not listening to me, so if someone could trigger another test run, I'd appreciate that.

vanzin commented on bin/spark-shell:
```diff
@@ -35,7 +35,8 @@ export _SPARK_CMD_USAGE="Usage: ./bin/spark-shell [options]"
 # do this specifically for the Spark shell because the scala REPL
 # has its own class loader, and any additional classpath specified
-# through spark.driver.extraClassPath is not automatically propagated.
+# through spark.driver.extraClassPath and/or spark.common.extraClassPath are not
```
I'm not sure how accurate this comment is anymore.

vanzin replied:
(Nevermind, the option is to enable those spark options to be respected by the shell.)

vanzin commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -465,6 +466,10 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
           |Please instead use:
           | - ./spark-submit with --driver-class-path to augment the driver classpath
           | - spark.executor.extraClassPath to augment the executor classpath
+          | - spark.common.extraClassPath to augument the driver and executor classpath.
```
augment

vanzin commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -505,6 +510,23 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
     getAll.sorted.map{case (k, v) => k + "=" + v}.mkString("\n")
   }
 
+  def getDriverExtraClassPath() : Option[String] = {
```
These should not be public; the user app shouldn't need to know about these.

vanzin commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -505,6 +510,23 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
+    val commonExtraClassPath = getOption("spark.common.extraClassPath")
+    val driverExtraClassPath = getOption("spark.driver.extraClassPath")
+    Some((commonExtraClassPath ++ driverExtraClassPath).mkString(File.pathSeparator))
+      .filter(_.nonEmpty)
```
To be pedantic, the filtering should be done earlier, otherwise:

    scala> Seq("", "").mkString(":")
    res0: String = :

vanzin commented on core/src/main/scala/org/apache/spark/deploy/Client.scala:
```diff
@@ -67,8 +67,15 @@ private class ClientEndpoint(
+        val commonClassPathConf = "spark.common.extraClassPath"
+
+        val driverClassPath = Some(
+          (sys.props.get(commonClassPathConf) ++ sys.props.get(driverClassPathConf))
```
Not your fault; but it's slightly weird that this code is using `sys.props` instead of the `SparkConf` object provided in the constructor.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:
```diff
@@ -624,3 +639,15 @@ object UserClasspathFirstTest {
+    val resource = ccl.getResourceAsStream("test.resource")
+    val bytes = ByteStreams.toByteArray(resource)
+    val contents = new String(bytes, 0, bytes.length, UTF_8)
+    if (contents != "COMMON") {
```
minor, but you could just use `assert` here.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:
```diff
@@ -143,6 +144,9 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
       .orElse(sparkProperties.get("spark.master"))
       .orElse(env.get("MASTER"))
       .orNull
+    commonExtraClassPath = Option(commonExtraClassPath)
```
The change in this file is not technically necessary, unless you add a specific command line option for this config (which I don't think you should). See how it doesn't have a field for the executor's extra class path (because there's no command line option for it).

vanzin commented:
Looks sane but there's some room for cleanup; also, need to resolve conflicts.

andrewor14 commented:
I'm not super convinced that we should add something like this. This patch introduces a config that's not really necessary and has potentially unclear semantics. E.g. What happens if I specify different versions of some jar on both common and executor class path? Which one gets loaded first? What if I pass in `--common-class-path` and set `spark.executor.extraClassPath` at the same time, is it the same?

Either way this patch has gone stale over time and has not been active for more than 2 months. I would recommend that we close this patch for now and re-open it later if there is more interest.

markgrover commented:
Hey Andrew, thanks for your feedback.

> What happens if I specify different versions of some jar on both common and executor class path? Which one gets loaded first? What if I pass in --common-class-path and set spark.executor.extraClassPath at the same time, is it the same?

The semantics are that common classpath is always added before the individual driver/executor classpath. I can clarify that in the documentation.

As far as your other point about two versions of the same jar goes, one could do that just by using spark.[driver|executor].extraClassPath too and the same logic would apply.

I was actually planning on updating this patch this week before the holidays roll in so I'd definitely vouch for not closing this. My motivation here is that many spark users, want to update both their driver and executor classpaths and either forget, miss or find cumbersome to update 2 different places for doing so. Common classpath would be one single go-to place for such users.

Would you be open to taking another look at this once I update this patch this week?

andrewor14 commented:
Sure, I wasn't sure if this would be updated any time soon.

tgravescs commented:
I'm a bit on the fence about this.  Its kind of bringing us full circle to what SPARK used to have when they had env variables, most of those applied across everything.  I think we chose to do them separate so users had the option of setting different ones.   I see the case for this makes it easier for users in some cases but I also see it as possibly confusing and just more maintenance and complexity in the spark code. My other concern here is now this has different semantics then extraJavaOptions and extraLibraryPath as they won't have a common option.

cc @pwendell as I think he originally added the configs to see if he remember any discussion around common ones.

andrewor14 commented:
> I see the case for this makes it easier for users in some cases but I also see it as possibly confusing and just more maintenance and complexity in the spark code.

That was exactly my concern as well. We had to go through the pain of deprecating most of those overly general environment variables but still keeping them around for backward compatibility.

vanzin commented:
I actually haven't seen a case of users needing different class paths for the driver and executor; in general, the executor class path is either the same or a subset of the driver's, so that using the same in both places leads to the desired behavior.

On the flip side, I've seen many people try to change just the driver class path and still run into errors because they needed the same change on the executor's configuration.

srowen commented:
Eh ... I didn't see much of a resolution on this. Does anyone feel weakly, moderately, strongly for/against this?

vanzin commented:
I still think this is a good idea; just need to clearly define the semantics (in case they aren't). e.g. `driver.extraClassPath` should be added before `common.extraClassPath` so it's possible to override classes in the common config.

markgrover commented:
Thanks for bringing this up, Sean. Yeah, I'd like to see this go in too. I realize there are some merge conflicts here. I will make the semantics clear in documentation (they are already there, will make sure they are clear enough).

So, if someone is strongly against, let me know. Otherwise, I will be updated the code early next week.

tgravescs commented:
I have seen cases where driver and executor needs different classpath, mostly in client mode.  gateways/launcher boxes can have different setup then hadoop nodes.  I have seen other people in jira/mailing lists mention this also.

I think I still have the same concerns.  Is the benefit here over possible user confusion worth it?  If the semantics are all clearly defined and its clear to the user and you can see this helping a bunch of your users then I'm OK with it.

SparkQA commented:
[Test build #66133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66133/consoleFull) for   PR 7739 at commit [`bca4fa3`](https://github.com/apache/spark/commit/bca4fa303a6de1b8b395c9fc6c2357edb9e12d5e).

SparkQA commented:
[Test build #66133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66133/console) for   PR 7739 at commit [`bca4fa3`](https://github.com/apache/spark/commit/bca4fa303a6de1b8b395c9fc6c2357edb9e12d5e).
 * This patch **fails SparkR unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66133/
Test FAILed.
